### PR TITLE
FastALS method

### PR DESCRIPTION
**Pull request by alexmul1114**
Continued from #38.

### CODE REVIEW

**Comment by alexmul1114** [comment]
Looks like there is probably minimal difference between using hcat and stack:
# Benchmark Report for `GCPDecompositions`

## Job Properties

<details>

* Time of benchmarks:
    - Target: 23 Feb 2024 - 14:59
    - Baseline: 23 Feb 2024 - 15:04
* Package commits:
    - Target: cf7fa6
    - Baseline: 47a06e
* Julia commits:
    - Target: 312098
    - Baseline: 312098
* Julia command flags:
    - Target: None
    - Baseline: None
* Environment variables:
    - Target: `GCP_BENCHMARK_SUITES => leastsquares`
    - Baseline: `GCP_BENCHMARK_SUITES => leastsquares`

</details>


## Results

<details>

A ratio greater than `1.0` denotes a possible regression (marked with :x:), while a ratio less
than `1.0` denotes a possible improvement (marked with :white_check_mark:). Only significant results - results
that indicate possible regressions or improvements - are shown below (thus, an empty table means that all
benchmark results remained invariant between builds).

| ID                                                                           | time ratio                   | memory ratio |
|------------------------------------------------------------------------------|------------------------------|--------------|
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=1"]`      |                1.28 (5%) :x: |   1.00 (1%)  |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=10"]`     |                1.05 (5%) :x: |   1.00 (1%)  |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=50"]`     |                1.13 (5%) :x: |   0.99 (1%)  |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=1"]`                        |                1.24 (5%) :x: |   1.01 (1%)  |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=10"]`                       | 0.73 (5%) :white_check_mark: |   1.00 (1%)  |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=50"]`                       | 0.87 (5%) :white_check_mark: |   1.00 (1%)  |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=1"]`                        |                1.75 (5%) :x: |   1.00 (1%)  |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=10"]`                       | 0.67 (5%) :white_check_mark: |   1.00 (1%)  |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=50"]`                       |                1.75 (5%) :x: |   1.00 (1%)  |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=1"]`                        |                1.18 (5%) :x: |   1.00 (1%)  |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=10"]`                       |                1.06 (5%) :x: |   1.00 (1%)  |

</details>


## Benchmark Group List

<details>

Here's a list of all the benchmark groups executed by this job:

- `["leastsquares"]`

</details>


## Julia versioninfo


<details>

### Target
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz   10622009            0      5760461    1393113133       190103  ticks
  Memory: 31.726390838623047 GB (14341.58984375 MB free)
  Uptime: 182098.406 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

### Baseline
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz   10882853            0      5871399    1397152068       196039  ticks
  Memory: 31.726390838623047 GB (14281.48828125 MB free)
  Uptime: 182374.078 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

</details>

**Comment by alexmul1114** [comment]
Added use of buffers for saved data between modes, here is the comparison against the previous FastALS implementation:
# Benchmark Report for `GCPDecompositions`

## Job Properties

<details>

* Time of benchmarks:
    - Target: 14 Mar 2024 - 08:29
    - Baseline: 14 Mar 2024 - 08:32
* Package commits:
    - Target: a6bd46
    - Baseline: f5182f
* Julia commits:
    - Target: 312098
    - Baseline: 312098
* Julia command flags:
    - Target: None
    - Baseline: None
* Environment variables:
    - Target: `GCP_BENCHMARK_SUITES => leastsquares`
    - Baseline: `GCP_BENCHMARK_SUITES => leastsquares`

</details>


## Results

<details>

A ratio greater than `1.0` denotes a possible regression (marked with :x:), while a ratio less
than `1.0` denotes a possible improvement (marked with :white_check_mark:). Only significant results - results
that indicate possible regressions or improvements - are shown below (thus, an empty table means that all
benchmark results remained invariant between builds).

| ID                                                                           | time ratio                   | memory ratio                 |
|------------------------------------------------------------------------------|------------------------------|------------------------------|
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=1"]`      | 0.85 (5%) :white_check_mark: | 0.58 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=10"]`     |                1.09 (5%) :x: | 0.51 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=50"]`     |                1.18 (5%) :x: | 0.58 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=1"]`  |                   0.99 (5%)  | 0.40 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=10"]` |                1.09 (5%) :x: | 0.38 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=50"]` | 0.91 (5%) :white_check_mark: | 0.41 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=1"]`  |                1.32 (5%) :x: | 0.68 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=10"]` | 0.94 (5%) :white_check_mark: | 0.36 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=50"]` | 0.88 (5%) :white_check_mark: | 0.37 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=1"]`      | 0.64 (5%) :white_check_mark: | 0.50 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=10"]`     | 0.58 (5%) :white_check_mark: | 0.47 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=50"]`     |                1.16 (5%) :x: | 0.50 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=1"]`                        | 0.86 (5%) :white_check_mark: | 0.59 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=10"]`                       | 0.80 (5%) :white_check_mark: | 0.46 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=50"]`                       |                   1.01 (5%)  | 0.57 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=1"]`                        | 0.77 (5%) :white_check_mark: | 0.45 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=10"]`                       | 0.88 (5%) :white_check_mark: | 0.39 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=50"]`                       | 0.90 (5%) :white_check_mark: | 0.43 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=1"]`                        | 0.78 (5%) :white_check_mark: | 0.38 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=10"]`                       | 0.80 (5%) :white_check_mark: | 0.36 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=50"]`                       | 0.90 (5%) :white_check_mark: | 0.37 (1%) :white_check_mark: |

</details>


## Benchmark Group List

<details>

Here's a list of all the benchmark groups executed by this job:

- `["leastsquares"]`

</details>


## Julia versioninfo


<details>

### Target
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz  115790899            0     86968679    15205469148      2397432  ticks
  Memory: 31.726390838623047 GB (9448.2734375 MB free)
  Uptime: 1.883099078e6 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

### Baseline
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz  115897541            0     87008212    15207638227      2398041  ticks
  Memory: 31.726390838623047 GB (9670.80078125 MB free)
  Uptime: 1.883243781e6 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

</details>

**Comment by alexmul1114** [comment]
Benchmark Report after adding buffers for all of the khatri-rao products. Getting a good speedup and memory improvement for the order 4 and 5 cases, but still using a little extra memory for the order 3 cases.
# Benchmark Report for `GCPDecompositions`

## Job Properties

<details>

* Time of benchmarks:
    - Target: 14 Mar 2024 - 20:31
    - Baseline: 14 Mar 2024 - 20:34
* Package commits:
    - Target: 18f857
    - Baseline: 7cb16a
* Julia commits:
    - Target: 312098
    - Baseline: 312098
* Julia command flags:
    - Target: None
    - Baseline: None
* Environment variables:
    - Target: `GCP_BENCHMARK_SUITES => leastsquares`
    - Baseline: `GCP_BENCHMARK_SUITES => leastsquares`

</details>


## Results

<details>

A ratio greater than `1.0` denotes a possible regression (marked with :x:), while a ratio less
than `1.0` denotes a possible improvement (marked with :white_check_mark:). Only significant results - results
that indicate possible regressions or improvements - are shown below (thus, an empty table means that all
benchmark results remained invariant between builds).

| ID                                                                           | time ratio                   | memory ratio                 |
|------------------------------------------------------------------------------|------------------------------|------------------------------|
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=1"]`      | 0.38 (5%) :white_check_mark: | 0.38 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=10"]`     | 0.28 (5%) :white_check_mark: | 0.33 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=50"]`     | 0.11 (5%) :white_check_mark: | 0.32 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=1"]`  | 0.32 (5%) :white_check_mark: | 0.18 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=10"]` | 0.17 (5%) :white_check_mark: | 0.16 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=50"]` | 0.13 (5%) :white_check_mark: | 0.28 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=1"]`  | 0.39 (5%) :white_check_mark: | 0.04 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=10"]` | 0.25 (5%) :white_check_mark: | 0.19 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=50"]` | 0.22 (5%) :white_check_mark: | 0.26 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=1"]`      | 0.33 (5%) :white_check_mark: | 0.17 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=10"]`     | 0.18 (5%) :white_check_mark: | 0.12 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=50"]`     | 0.17 (5%) :white_check_mark: | 0.25 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=1"]`                        | 0.82 (5%) :white_check_mark: | 0.94 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=10"]`                       |                   0.97 (5%)  |                1.26 (1%) :x: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=50"]`                       |                   0.97 (5%)  |                1.10 (1%) :x: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=1"]`                        | 0.58 (5%) :white_check_mark: | 0.99 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=10"]`                       | 0.70 (5%) :white_check_mark: |                1.41 (1%) :x: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=50"]`                       | 0.94 (5%) :white_check_mark: |                1.16 (1%) :x: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=1"]`                        | 0.46 (5%) :white_check_mark: |                1.06 (1%) :x: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=10"]`                       | 0.47 (5%) :white_check_mark: |                1.50 (1%) :x: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=50"]`                       | 0.73 (5%) :white_check_mark: |                1.22 (1%) :x: |

</details>


## Benchmark Group List

<details>

Here's a list of all the benchmark groups executed by this job:

- `["leastsquares"]`

</details>


## Julia versioninfo


<details>

### Target
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz  118963244            0     89680258    15625943959      2490480  ticks
  Memory: 31.726390838623047 GB (10910.5078125 MB free)
  Uptime: 1.926389953e6 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

### Baseline
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz  119260867            0     89742947    15629176119      2491945  ticks
  Memory: 31.726390838623047 GB (10286.14453125 MB free)
  Uptime: 1.926614484e6 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

</details>

**Comment by alexmul1114** [comment]
Modified the helper function to use views more, now using less memory for all cases.
# Benchmark Report for `GCPDecompositions`

## Job Properties

<details>

* Time of benchmarks:
    - Target: 15 Mar 2024 - 08:45
    - Baseline: 15 Mar 2024 - 08:50
* Package commits:
    - Target: 6a163e
    - Baseline: 7cb16a
* Julia commits:
    - Target: 312098
    - Baseline: 312098
* Julia command flags:
    - Target: None
    - Baseline: None
* Environment variables:
    - Target: `GCP_BENCHMARK_SUITES => leastsquares`
    - Baseline: `GCP_BENCHMARK_SUITES => leastsquares`

</details>


## Results

<details>

A ratio greater than `1.0` denotes a possible regression (marked with :x:), while a ratio less
than `1.0` denotes a possible improvement (marked with :white_check_mark:). Only significant results - results
that indicate possible regressions or improvements - are shown below (thus, an empty table means that all
benchmark results remained invariant between builds).

| ID                                                                           | time ratio                   | memory ratio                 |
|------------------------------------------------------------------------------|------------------------------|------------------------------|
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=1"]`      | 0.41 (5%) :white_check_mark: | 0.32 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=10"]`     | 0.29 (5%) :white_check_mark: | 0.23 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30), rank(X)=50"]`     | 0.89 (5%) :white_check_mark: | 0.47 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=1"]`  | 0.32 (5%) :white_check_mark: | 0.14 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=10"]` | 0.31 (5%) :white_check_mark: | 0.10 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(15, 20, 25, 30, 35), rank(X)=50"]` | 0.35 (5%) :white_check_mark: | 0.23 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=1"]`  | 0.40 (5%) :white_check_mark: | 0.37 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=10"]` | 0.22 (5%) :white_check_mark: | 0.11 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 30, 30, 30, 30), rank(X)=50"]` | 0.30 (5%) :white_check_mark: | 0.19 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=1"]`      | 0.48 (5%) :white_check_mark: | 0.13 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=10"]`     | 0.25 (5%) :white_check_mark: | 0.07 (1%) :white_check_mark: |
| `["leastsquares", "least-squares-size(X)=(30, 40, 50, 60), rank(X)=50"]`     | 0.47 (5%) :white_check_mark: | 0.21 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=1"]`                        | 0.78 (5%) :white_check_mark: | 0.84 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=10"]`                       | 0.85 (5%) :white_check_mark: | 0.93 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(15, 20, 25), rank(X)=50"]`                       | 0.87 (5%) :white_check_mark: |                   0.99 (1%)  |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=1"]`                        | 0.57 (5%) :white_check_mark: | 0.83 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=10"]`                       | 0.70 (5%) :white_check_mark: | 0.89 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(30, 40, 50), rank(X)=50"]`                       | 0.74 (5%) :white_check_mark: | 0.98 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=1"]`                        | 0.42 (5%) :white_check_mark: | 0.81 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=10"]`                       |                1.24 (5%) :x: | 0.84 (1%) :white_check_mark: |
| `["leastsquares", "size(X)=(60, 70, 80), rank(X)=50"]`                       | 0.70 (5%) :white_check_mark: | 0.95 (1%) :white_check_mark: |

</details>


## Benchmark Group List

<details>

Here's a list of all the benchmark groups executed by this job:

- `["leastsquares"]`

</details>


## Julia versioninfo


<details>

### Target
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz  120025868            0     90644744    15910842369      2513869  ticks
  Memory: 31.726390838623047 GB (10260.5546875 MB free)
  Uptime: 1.97043814e6 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

### Baseline
```
Julia Version 1.10.0
Commit 3120989f39 (2023-12-25 18:01 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
      Microsoft Windows [Version 10.0.22621.3155]
  CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz: 
                 speed         user         nice          sys         idle          irq
       #1-16  2304 MHz  120422368            0     90844071    15915169694      2521088  ticks
  Memory: 31.726390838623047 GB (10327.34765625 MB free)
  Uptime: 1.970745843e6 sec
  Load Avg:  0.0  0.0  0.0
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, tigerlake)
  Threads: 1 on 16 virtual cores
```

</details>

File: src/gcp-algorithms/fastals.jl [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+    # Initialization
+    M = deepcopy(init)
+
+    # Determine order of modes of MTTKRP to compute
+    Jns = [prod(size(X)[1:n]) for n in 1:N]
+    Kns = [prod(size(X)[n+1:end]) for n in 1:N]
+    Kn_minus_ones = [prod(size(X)[n:end]) for n in 1:N]
+    comp = Jns .<= Kn_minus_ones
+    n_star = maximum(map(x -> comp[x] ? x : 0, 1:N))
```
Comment (dahong67) [inline src/gcp-algorithms/fastals.jl:36]: Is this what these lines are trying to do?
```suggestion
    n_star = findlast(n -> Jns[n] <= Kn_minus_ones[n], 1:N)
```

**Review by dahong67 — COMMENTED** [review]
Please go through the requested changes.

I think we can also simplify the code a decent amount, will look more later!

File: src/gcp-algorithms/fastals.jl [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    buffers = create_FastALS_buffers(M.U, order, Jns, Kns)
+
+    for _ in 1:algorithm.maxiters
+        FastALS_iter!(X, M, order, Jns, Kns, buffers)
+    end
+
+    return M
+end
+
+"""
```
[the remark this answers is not in the document]
Comment (dahong67) [reply src/gcp-algorithms/fastals.jl:51]: Addressing this together during our meeting, marking resolved.

**Comment by dahong67** [comment]
Looking good! Merging. 🎊

**Pull request merged**